torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -597,29 +597,38 @@ void CudaCodeGen::call(const std::vector<CallArg>& args) {
   USE_TRIGGER(cuda_codegen_executed);
 }
 
+void CudaSetContext(CUcontext pctx) {
+  if (!pctx) {
+    std::unique_lock<std::mutex> cudaFreeMutexLock(
+        *(c10::cuda::CUDACachingAllocator::getFreeMutex()));
+    cudaFree(0);
```

Comment:
What is `cudaFree(0);` doing here?

Reply:
In fact I'm sort of confused by the existence of this function, what does it do?

Reply:
This is needed to pick up the new device. As XQ mentioned 
One is at Cuda Runtime API level: cudaSetDevice, and the other is at Cuda Driver API level: cuModuleLoad. The cudaSetDevice does not have to really change the underlying device if it doesn't have to. cudaFree forces it to make that change.